### PR TITLE
[Merged by Bors] - feat(meta): add primitives for GPT integration

### PR DESCRIPTION
**Pull request by EdAyers**
New primitives:
- `environment.mark_namespace_as_open`
- `environment.execute_open`
- `environment.get_namespaces`
- `expr.is_string_macro`
- `lean.parser.run`

Coauthor: @jesse-michael-han 

These changes are needed to support the GPT-f suggestions system.

### CODE REVIEW

**Comment by EdAyers** [comment]
Currently the test breaks because for some reason when the parser is initialised it doesn't pick up the passed string.

**Comment by EdAyers** [comment]
A test is failing:
```lean
run_cmd do {ps ← lean.parser.mk_parser_state, tactic.trace ps.cur_pos}
```

this code is nondeterministic. I've observed it producing `⟨1,0⟩` or `⟨32641, 1932892776⟩` or `⟨0,2⟩` :-/
I guess somewhere a dead pointer is being dereferenced.

**Comment by EdAyers** [comment]
Ok I think that the scanner does not initialise position information until scanning has started, so if you immediately read `scanner.m_line` you just get stack rubble. So I am going to remove this test.

File: library/init/meta/lean/parser.lean [hunk 1]
```diff
@@ -150,5 +152,13 @@ end reflectable
...
 r.expr
 
+meta constant run {α} : parser α → tactic α
+
+meta def run_with_input {α} : parser α → string → tactic α := λ p s,
+prod.fst <$> (run (with_input p s) )
+
+meta def mk_parser_state : tactic lean.parser_state :=
+lean.parser.run_with_input lean.parser.get_state ""
+
```
[the remark this answers is not in the document]
Comment (jesse-michael-han) [reply library/init/meta/lean/parser.lean]: Thanks for the feedback Gabriel! Removed as requested.

**Pull request closed without merging**